File: java/src/com/algorithims/sorting/HeapSort.java
package com.algorithims.sorting;

import com.utility.Utility;

/**
 * 
 * @author saurabh_mahambrey HEAP SORT: Since a Binary Heap is a Complete Binary
 *         Tree, it can be easily represented as array and array based
 *         representation is space efficient. If the parent node is stored at
 *         index I, the left child can be calculated by 2 * I + 1 and right
 *         child by 2 * I + 2 (assuming the indexing starts at 0).
 * 
 *         Heap Sort Algorithm for sorting in increasing order: 1. Build a max
 *         heap from the input data. 2. At this point, the largest item is
 *         stored at the root of the heap. Replace it with the last item of the
 *         heap followed by reducing the size of heap by 1. Finally, heapify the
 *         root of tree. 3. Repeat above steps while size of heap is greater
 *         than 1.
 * 
 *         time complexity of Heap Sort is O(nLogn)
 */
public class HeapSort {

	public void sort(int arr[]) {
		int n = arr.length;
		for (int i = n / 2 - 1; i >= 0; i--) {
			heapify(arr, n, i);
		}

		for (int i = n - 1; i > 0; i--) {
			Utility.swapElements(arr, 0, i);

			heapify(arr, i, 0);
		}

	}

	private void heapify(int[] arr, int n, int i) {
		// TODO Auto-generated method stub
		int largest = i;
		int left = i * 2 + 1;
		int right = i * 2 + 2;

		if (left < n && arr[largest] < arr[left]) {
			largest = left;
		}
		if (right < n && arr[largest] < arr[right]) {
			largest = right;
		}
		if (largest != i) {
			Utility.swapElements(arr, largest, i);
			heapify(arr, n, largest);
		}

	}

	public static void main(String[] args) {
		HeapSort obj = new HeapSort();
		int[] arr = { 1, 22, 221, 5, 55, 40, 42, 33, 1, 2, 2323, 1, 23, 44 };
		obj.sort(arr);
		Utility.printArray(arr);
	}
}

File: python/src/algorithims/sort/heap_sort.py
"""
Problem Statement: Sort array using heap sort algorithm
Time Complexity = 0(nlog(n))
Link: https://www.geeksforgeeks.org/heap-sort/
"""

"""
Algorithm:
    heapSort:
        1. Build a max heap from the input data.
        2. At this point, the largest item is stored at the root of the heap. Replace it with the last item of the heap, 
           followed by reducing the size of heap by 1. Finally, heapify the root of tree.
        3. Repeat above steps while size of heap is greater than 1.    
"""


def heapify(in_arr, i, n):
    """
    creates a max heap from the recursive array
    :param in_arr: input array whose heap is to be created
    :param i: current parent element
    :param n: length of the array
    :return: void
    """
    l_child = 2 * i + 1
    r_child = 2 * i + 2
    max_ele = i
    if l_child < n and in_arr[l_child] > in_arr[max_ele]:
        max_ele = l_child
    if r_child < n and in_arr[r_child] > in_arr[max_ele]:
        max_ele = r_child
    if max_ele != i:
        in_arr[i], in_arr[max_ele] = in_arr[max_ele], in_arr[i]
        heapify(in_arr, max_ele, n)


def heap_sort(input_arr):
    """
    sorts the input array using heap sort method
    :param input_arr: array to be sorted
    """
    n = len(input_arr)
    for i in range(int(n/2) - 1, -1, -1):
        heapify(input_arr, i, n)
    print("heap done")
    print(input_arr)
    for i in range(1, n):
        input_arr[0], input_arr[n - i] = input_arr[n - i], input_arr[0]
        heapify(input_arr, 0, n - i)

    print(input_arr)


"""
Test Cases:
"""

if __name__ == '__main__':
    test_arr = [22, 1, 55, 43, 12, -6]
    heap_sort(test_arr)

